Feature/ci run boardfarm

rmelotte opened the pull request:
Run boardfarm using docker-only in CI, until the docker-compose integration can be done.

rmelotte committed:
docker: add a Dockerfile containing boardfarm

To run the boardfarm tests, we need an image containing boardfarm.

Add an image that contains all the boardfarm dependencies, fetch a
specific commit (the latest one from the master branch) and install
it.

The dockerfile was adapted from the existing work on:
https://github.com/prplfoundation/prplMesh/pull/1483

Also add the image to .gitlab-ci to get it built automatically.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tests: boardfarm: use the installed version of boardfarm

To use an already-installed boardfarm version we don't need to know
it's path, we can assume it's already in PATH.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
ci: add a job to run the boardfarm tests

Because prplMesh is not currently relocatable, we have to run it both
in the same directory it was built in, but also as the same user.
It is currently built with the root user inside a docker container,
with the USER variable set as ''.

As a consequence, we have to run it as root as well, and explicitly
set USER to an empty string (because boardfarm relies on it's
presence).

Since we want to run boardfarm in docker and boardfarm starts
containers itself, we have to use a docker image where docker itself
is available from within the container.

For this 3 options were possible:

- The current one, running another docker daemon inside the first
docker container. It is a little bit overkill, and makes it hard to
preserve the cache between runs.

- Run with a Docker executor and bind-mount the Docker socket. It
requires some care when we want to mount volumes from within a
container, because the volumes have to be available on the host.

- Run with a shell executor. This approach has the benefit of always
having the prplMesh tree cloned and the artifacts available on the
host. Since we currently run the containers as root however, the
Gitlab runner won't be able to clean up files in-between runs.

The option 2 was chosen, because it was the easiest to use at the
moment.

For now the job also needs only the docker builds, so add a "need" for
it.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tests: boardfarm: error out and print the result when pexpect run fails

When a command failed to run, raise an exception early, and add the
stdout and stderr output of the command in the message.

These changes were taken and adapted from:
https://github.com/prplfoundation/prplMesh/pull/1483

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

arnout commented on tools/docker/boardfarm-ci/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+#RUN pip3 install jsonschema distro
+
+
+RUN curl -fsSL https://download.docker.com/linux/debian/gpg | apt-key add -
+RUN add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable"
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+    docker-ce \
+    docker-ce-cli \
+    containerd.io \
+    && rm -rf /var/lib/apt/lists/*
+
+RUN curl -L "https://github.com/docker/compose/releases/download/1.26.0/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
+RUN chmod 755 /usr/local/bin/docker-compose
```
For my info: why isn't all this simply taken from Debian/Ubuntu repos? Are we using such advanced features?

rmelotte replied:
Good question, I took the dockerfile from https://github.com/prplfoundation/prplMesh/pull/1483 and tried to make only the changes I needed.
Maybe @odkq needs a more recent version than the base image provides?

odkq replied:
We could, debian stable is 1.21 and I don't think we are going to use later development, but as documentation is not tagged (always the latest) with this we have the mechanism in place for future upgrades (just change the 1.26.0 in the url and test)

arnout commented on tools/docker/boardfarm-ci/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+COPY requirements.txt /app/requirements.txt
+
+WORKDIR app
+RUN pip3 install -r requirements.txt
```
Why do we need this? Before, we just took the requirements.txt from boardfarm. I.e.
```

RUN git clone https://github.com/mattsm/boardfarm.git \
    && cd boardfarm \
    && git checkout 100521fde1fb67536682cafecc2f91a6e2e8a6f8 \
    && pip3 install -r requirements.txt
    && python3 setup.py install 
```

Yes, it does mean that all the pip stuff has to be re-downloaded whenever we change the boardfarm hash, but that's going to be rare anyway.

odkq replied:
Boardfarm itself is not installed in this setup, it is used from the rootdir/boardfarm dir directly to allow for any changes made locally to be used.

As we do not expect to change boardfarm itself, this will be changed to installing as before, cloning and the rest.

rmelotte replied:
@arnout good point, I hadn't noticed boardfarm itself also included the requirements, I though it was added by us.

@odkq with these changes, boardfarm _is_ installed in the image itself. Is it something you want to do differently?

arnout commented on tests/run_bf.sh:
```diff
@@ -15,4 +15,4 @@ else
 export PYTHONPATH
 
-exec python3 "${bf_dir}"/bft -c "${bf_plugins_dir}"/boardfarm_prplmesh/prplmesh_config.json -n prplmesh_docker -x test_flows
+exec bft -c "${bf_plugins_dir}"/boardfarm_prplmesh/prplmesh_config.json -n prplmesh_docker -x test_flows
```
When Oleksii originally integrated boardfarm, I asked *not* do to this exactly because it requires installing boardfarm, which I wanted to avoid :-)

But I guess it's OK now that we have it in docker.

odkq replied:
This change is actually bogus; I added a script run_bf_compose.sh that still uses the boardfarm in it's local copy, as above

arnout commented on .gitlab-ci.yml:
```diff
@@ -215,6 +215,20 @@ run-tests:
+  needs:
+    - job: build-in-docker
+  tags:
+    - boardfarm
```
The commit message says something completely different...

> It runs in a shell runner

No, it doesn't have the shell tag and it specifies an image.

> It also requires docker to be available on the runner, so add the
"docker-build" tag.

No, it doesn't have the docker-build tag.

rmelotte replied:
I changed it back and forth when I noticed the shell runner had the issue with the user permissions.
It's a docker executor now, I edited the commit message.

arnout commented on .gitlab-ci.yml:
```diff
@@ -215,6 +215,20 @@ run-tests:
+  stage: test
+  image: $CI_REGISTRY_IMAGE/prplmesh-boardfarm-ci:$CI_PIPELINE_ID
+  script:
+    - export USER=gitlab-runner BFT_DEBUG=y
```
These exports should be explained in a comment or in the commit message.

rmelotte replied:
Done.

rmelotte commented:
Some of the commits belongs to @odkq , so I still need to add a SOB line for them.

arnout reviewed (approved):
A few of my comments haven't been addressed yet, but it's anyway stuff that can be done later, so approving already.

abelog commented on .gitlab-ci.yml:
```diff
@@ -212,6 +215,20 @@ run-tests:
+    - tests/run_bf.sh
+  artifacts:
+    paths:
+      - logs
```
Boardfarm stores it's logs in `results` folder, we should store it also. It have all logs received from consoles of devices during tests and some miscellaneous info.
```suggestion
      - logs
      - results
```

rmelotte replied:
Indeed, thanks for the suggestion I applied it.

abelog commented on .gitlab-ci.yml:
```diff
@@ -212,6 +215,20 @@ run-tests:
+  stage: test
+  image: $CI_REGISTRY_IMAGE/prplmesh-boardfarm-ci:$CI_PIPELINE_ID
+  script:
+    - export USER=''
```
Why `Null`-fy `$USER` variable?. Internally, in `PrplMeshDocker` it's used to generate `docker network` ID.

rmelotte replied:
Currently prplMesh is built as root, and also has to be run as root with and empty USER value (see https://jira.prplfoundation.org/projects/PPM/issues/PPM-208 and https://jira.prplfoundation.org/projects/PPM/issues/PPM-209).
Because of that we also need to run boardfarm with the same (empty) user.

abelog reviewed (approved):
In general, LGTM.
Small questions left, but nothing big.

rmelotte committed:
tests: boardfarm: use docker_name for check_status

With https://github.com/prplfoundation/prplMesh/pull/1486, "name" was
changed to "docker_name".

The name in "check_status" was not updated, so do it now.

It fixes the following error:
Error: No such container: controller

See for example:
https://gitlab.com/prpl-foundation/prplMesh/-/jobs/625370585

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tests: boardfarm: use device_get_info for check_status

With https://github.com/prplfoundation/prplMesh/pull/1486, the first
device to start has been changed to be the agent.

Since the agent cannot be operational unless there is a controller
already running on the same network, something else has to be used for
"check_status".

For now, use netcat to check that "device_get_info" works.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>
